File: .tools/html2pdf.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

(async () => {
    const browser = await puppeteer.launch({ args: ['--no-sandbox'] });
    const files = process.argv[2] ? [process.argv[2]] : [];

    if (!files.length) {
        // 递归查找 public/html 目录下的所有 .html 文件
        function walk(dir) {
            let results = [];
            fs.readdirSync(dir).forEach(file => {
                const fullPath = path.join(dir, file);
                if (fs.statSync(fullPath).isDirectory()) {
                    results = results.concat(walk(fullPath));
                } else if (file.endsWith('.html')) {
                    results.push(fullPath);
                }
            });
            return results;
        }
        files.push(...walk('public/html'));
    }

    for (const htmlfile of files) {
        const rel = path.relative('public/html', htmlfile);
        const pdffile = path.join('public/pdf', rel.replace(/\.html$/, '.pdf'));
        fs.mkdirSync(path.dirname(pdffile), { recursive: true });
        const page = await browser.newPage();
        await page.goto('file://' + path.resolve(htmlfile), { waitUntil: 'networkidle0' });

        // 计算更精确的页面宽度
        const pageWidth = await page.evaluate(() => {
            const elements = document.querySelectorAll('body *');
            let max = document.body.scrollWidth;
            elements.forEach(el => {
                const rect = el.getBoundingClientRect();
                const right = rect.right + window.pageXOffset;
                if (right > max) {
                    max = right;
                }
            });
            return max;
        });

        // 定义一个固定的目标宽度和最小缩放比例
        const targetWidth = 1800;
        const minScale = 0.6;
        let scale = targetWidth / pageWidth;

        // 确保缩放比例不低于最小缩放比例
        if (scale < minScale) {
            scale = minScale;
        }

        await page.pdf({
            path: pdffile,
            width: `${targetWidth}px`,
            scale: scale,
            printBackground: true
        });

        await page.close();
        console.log(`Converted ${htmlfile} -> ${pdffile}`);
    }

    await browser.close();
})();

File: .tools/html2pdf.py
import os
import asyncio
from pyppeteer import launch
from pathlib import Path

async def convert_html_to_pdf(html_path, pdf_path):
    """将单个HTML文件转换为PDF"""
    browser = await launch(headless=True)
    page = await browser.newPage()
    
    # 读取HTML文件内容
    with open(html_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    # 设置视口为A4尺寸
    await page.setViewport({'width': 595, 'height': 842})
    
    # 加载HTML内容
    await page.goto(f'file://{os.path.abspath(html_path)}', waitUntil='networkidle0')
    
    # 计算内容的宽度和高度
    content_width = await page.evaluate('document.body.scrollWidth')
    content_height = await page.evaluate('document.body.scrollHeight')
    
    # 确定使用横向还是纵向
    is_landscape = content_width > content_height
    
    # 计算缩放比例
    width_scale = 595 / content_width
    height_scale = 842 / content_height
    scale = min(width_scale, height_scale)
    
    # 确保缩放比例不会过大
    scale = min(scale, 1.0)
    
    # 创建PDF保存目录
    os.makedirs(os.path.dirname(pdf_path), exist_ok=True)
    
    # 保存为PDF
    await page.pdf({
        'path': pdf_path,
        'format': 'A4',
        'landscape': is_landscape,
        'scale': scale,
        'printBackground': True
    })
    
    await browser.close()
    print(f"已转换: {html_path} -> {pdf_path}")

async def main():
    """主函数：处理所有HTML文件"""
    html_dir = Path('public/html')
    pdf_dir = Path('public/pdf')
    
    # 确保输出目录存在
    pdf_dir.mkdir(parents=True, exist_ok=True)
    
    # 获取所有HTML文件
    html_files = list(html_dir.rglob('*.html'))
    
    if not html_files:
        print("未找到HTML文件")
        return
    
    # 处理每个HTML文件
    for html_file in html_files:
        # 计算对应的PDF路径，保持目录结构
        relative_path = html_file.relative_to(html_dir)
        pdf_file = pdf_dir / relative_path.with_suffix('.pdf')
        
        # 确保父目录存在
        pdf_file.parent.mkdir(parents=True, exist_ok=True)
        
        # 转换HTML到PDF
        await convert_html_to_pdf(str(html_file), str(pdf_file))

if __name__ == "__main__":
    # 运行异步主函数
    asyncio.run(main())    